File: sdk/ts/src/securityCenter/alerts.ts
import type {
  SecurityCenterCustomerInsight,
  SecurityCenterVendorInsight,
} from "../clients/securityCenter.js";

export interface MonitoringEvent {
  id: string;
  entityType: "vendor" | "customer";
  entityId: string;
  eventType: "questionnaire" | "audit" | "incident" | "review";
  status: "open" | "resolved" | "acknowledged";
  occurredAt: Date;
  description: string;
  severity: "info" | "warning" | "critical";
}

export interface GovernanceAlert {
  id: string;
  entityType: "vendor" | "customer";
  entityId: string;
  title: string;
  detail: string;
  severity: "info" | "warning" | "critical";
  requiresReview: boolean;
  escalationTarget: string;
}

export interface MonitoringContext {
  vendorResolver?: (id: string) => SecurityCenterVendorInsight | undefined;
  customerResolver?: (id: string) => SecurityCenterCustomerInsight | undefined;
  escalationResolver?: (
    entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
    event: MonitoringEvent,
  ) => string;
}

export function evaluateMonitoringEvents(
  events: MonitoringEvent[],
  context: MonitoringContext,
): GovernanceAlert[] {
  const alerts: GovernanceAlert[] = [];

  for (const event of events) {
    const entity = resolveEntity(event, context);
    if (!entity) continue;

    switch (event.eventType) {
      case "incident":
        alerts.push(buildIncidentAlert(event, entity, context));
        break;
      case "audit":
        if (event.status !== "resolved") alerts.push(buildAuditAlert(event, entity, context));
        break;
      case "questionnaire":
        if (event.status === "open") alerts.push(buildQuestionnaireAlert(event, entity, context));
        break;
      case "review":
        if (isReviewOverdue(event)) alerts.push(buildReviewAlert(event, entity, context));
        break;
      default:
        alerts.push(buildGenericAlert(event, entity, context));
    }
  }

  return alerts;
}

function resolveEntity(
  event: MonitoringEvent,
  context: MonitoringContext,
): SecurityCenterVendorInsight | SecurityCenterCustomerInsight | undefined {
  if (event.entityType === "vendor") return context.vendorResolver?.(event.entityId);
  return context.customerResolver?.(event.entityId);
}

function buildIncidentAlert(
  event: MonitoringEvent,
  entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
  context: MonitoringContext,
): GovernanceAlert {
  return {
    id: `alert-${event.id}`,
    entityType: event.entityType,
    entityId: event.entityId,
    title: `Incident reported for ${entity.name}`,
    detail: event.description,
    severity: "critical",
    requiresReview: true,
    escalationTarget: resolveEscalation(entity, event, context, "grc-committee"),
  } satisfies GovernanceAlert;
}

function buildAuditAlert(
  event: MonitoringEvent,
  entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
  context: MonitoringContext,
): GovernanceAlert {
  return {
    id: `alert-${event.id}`,
    entityType: event.entityType,
    entityId: event.entityId,
    title: `Audit finding outstanding for ${entity.name}`,
    detail: event.description,
    severity: event.severity === "critical" ? "critical" : "warning",
    requiresReview: true,
    escalationTarget: resolveEscalation(entity, event, context, "audit-team"),
  } satisfies GovernanceAlert;
}

function buildQuestionnaireAlert(
  event: MonitoringEvent,
  entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
  context: MonitoringContext,
): GovernanceAlert {
  return {
    id: `alert-${event.id}`,
    entityType: event.entityType,
    entityId: event.entityId,
    title: `Questionnaire pending for ${entity.name}`,
    detail: event.description,
    severity: "warning",
    requiresReview: false,
    escalationTarget: resolveEscalation(entity, event, context, "compliance-team"),
  } satisfies GovernanceAlert;
}

function buildReviewAlert(
  event: MonitoringEvent,
  entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
  context: MonitoringContext,
): GovernanceAlert {
  return {
    id: `alert-${event.id}`,
    entityType: event.entityType,
    entityId: event.entityId,
    title: `Periodic review overdue for ${entity.name}`,
    detail: event.description,
    severity: "warning",
    requiresReview: true,
    escalationTarget: resolveEscalation(entity, event, context, "policy-owner"),
  } satisfies GovernanceAlert;
}

function buildGenericAlert(
  event: MonitoringEvent,
  entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
  context: MonitoringContext,
): GovernanceAlert {
  return {
    id: `alert-${event.id}`,
    entityType: event.entityType,
    entityId: event.entityId,
    title: `Monitoring event for ${entity.name}`,
    detail: event.description,
    severity: event.severity,
    requiresReview: event.severity !== "info",
    escalationTarget: resolveEscalation(entity, event, context, "governance"),
  } satisfies GovernanceAlert;
}

function resolveEscalation(
  entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
  event: MonitoringEvent,
  context: MonitoringContext,
  fallback: string,
): string {
  const target = context.escalationResolver?.(entity, event);
  return target ?? fallback;
}

function isReviewOverdue(event: MonitoringEvent): boolean {
  const now = Date.now();
  return now - event.occurredAt.getTime() > 14 * 24 * 60 * 60 * 1000;
}

File: src/cerebro_sdk/security_center/alerts.py
"""Monitoring alerts and governance escalation helpers."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Callable, Iterable, List, Optional

from .models import SecurityCenterCustomerInsight, SecurityCenterVendorInsight


@dataclass(slots=True)
class MonitoringEvent:
    id: str
    entity_type: str
    entity_id: str
    event_type: str
    status: str
    occurred_at: datetime
    description: str
    severity: str


@dataclass(slots=True)
class GovernanceAlert:
    id: str
    entity_type: str
    entity_id: str
    title: str
    detail: str
    severity: str
    requires_review: bool
    escalation_target: str


@dataclass(slots=True)
class MonitoringContext:
    vendor_resolver: Callable[[str], Optional[SecurityCenterVendorInsight]] | None = None
    customer_resolver: Callable[[str], Optional[SecurityCenterCustomerInsight]] | None = None
    escalation_resolver: Callable[
        [SecurityCenterVendorInsight | SecurityCenterCustomerInsight, MonitoringEvent],
        Optional[str],
    ] | None = None


def evaluate_monitoring_events(
    events: Iterable[MonitoringEvent],
    context: MonitoringContext,
) -> List[GovernanceAlert]:
    alerts: List[GovernanceAlert] = []
    for event in events:
        entity = _resolve_entity(event, context)
        if entity is None:
            continue

        if event.event_type == "incident":
            alerts.append(
                _build_alert(
                    event,
                    entity,
                    "Incident reported",
                    "critical",
                    requires_review=True,
                    fallback="grc-committee",
                    context=context,
                )
            )
        elif event.event_type == "audit" and event.status != "resolved":
            severity = "critical" if event.severity == "critical" else "warning"
            alerts.append(
                _build_alert(
                    event,
                    entity,
                    "Audit finding outstanding",
                    severity,
                    requires_review=True,
                    fallback="audit-team",
                    context=context,
                )
            )
        elif event.event_type == "questionnaire" and event.status == "open":
            alerts.append(
                _build_alert(
                    event,
                    entity,
                    "Questionnaire pending",
                    "warning",
                    requires_review=False,
                    fallback="compliance-team",
                    context=context,
                )
            )
        elif event.event_type == "review" and _is_review_overdue(event):
            alerts.append(
                _build_alert(
                    event,
                    entity,
                    "Periodic review overdue",
                    "warning",
                    requires_review=True,
                    fallback="policy-owner",
                    context=context,
                )
            )
        else:
            alerts.append(
                _build_alert(
                    event,
                    entity,
                    "Monitoring event",
                    event.severity,
                    requires_review=event.severity != "info",
                    fallback="governance",
                    context=context,
                )
            )

    return alerts


def _resolve_entity(
    event: MonitoringEvent,
    context: MonitoringContext,
) -> SecurityCenterVendorInsight | SecurityCenterCustomerInsight | None:
    if event.entity_type == "vendor":
        resolver = context.vendor_resolver
        return resolver(event.entity_id) if resolver else None
    resolver = context.customer_resolver
    return resolver(event.entity_id) if resolver else None


def _build_alert(
    event: MonitoringEvent,
    entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
    prefix: str,
    severity: str,
    *,
    requires_review: bool,
    fallback: str,
    context: MonitoringContext,
) -> GovernanceAlert:
    title = f"{prefix} for {entity.name}"
    escalation_target = _resolve_escalation(context, entity, event, fallback)
    return GovernanceAlert(
        id=f"alert-{event.id}",
        entity_type=event.entity_type,
        entity_id=event.entity_id,
        title=title,
        detail=event.description,
        severity=severity,
        requires_review=requires_review,
        escalation_target=escalation_target,
    )


def _resolve_escalation(
    context: MonitoringContext,
    entity: SecurityCenterVendorInsight | SecurityCenterCustomerInsight,
    event: MonitoringEvent,
    fallback: str,
) -> str:
    if context.escalation_resolver:
        target = context.escalation_resolver(entity, event)
        if target:
            return target
    return fallback


def _is_review_overdue(event: MonitoringEvent) -> bool:
    return datetime.utcnow() - event.occurred_at > timedelta(days=14)
